### PR TITLE
Mexc fetchFundingRateHistory

### PR DESCRIPTION
**Pull request by Dan-krm**
Created the fetchFundingRateHistory method for mexc.js
``` 
node examples/js/cli mexc fetchFundingRateHistory BTC_USDT
2021-12-11T00:44:40.804Z
Node.js: v16.9.1
CCXT v1.63.19
mexc.fetchFundingRateHistory (BTC_USDT)
486 ms
  symbol | fundingRate |     timestamp |                 datetime
-----------------------------------------------------------------
BTC_USDT |    0.000079 | 1638633600000 | 2021-12-04T16:00:00.000Z
BTC_USDT |    0.000132 | 1638662400000 | 2021-12-05T00:00:00.000Z
BTC_USDT |     0.00014 | 1638691200000 | 2021-12-05T08:00:00.000Z
BTC_USDT |     0.00015 | 1638720000000 | 2021-12-05T16:00:00.000Z
BTC_USDT |     0.00015 | 1638748800000 | 2021-12-06T00:00:00.000Z
BTC_USDT |     0.00015 | 1638777600000 | 2021-12-06T08:00:00.000Z
BTC_USDT |     0.00005 | 1638806400000 | 2021-12-06T16:00:00.000Z
BTC_USDT |     0.00015 | 1638835200000 | 2021-12-07T00:00:00.000Z
BTC_USDT |   -0.000011 | 1638864000000 | 2021-12-07T08:00:00.000Z
BTC_USDT |     0.00005 | 1638892800000 | 2021-12-07T16:00:00.000Z
BTC_USDT |     0.00015 | 1638921600000 | 2021-12-08T00:00:00.000Z
BTC_USDT |     0.00015 | 1638950400000 | 2021-12-08T08:00:00.000Z
BTC_USDT |     0.00005 | 1638979200000 | 2021-12-08T16:00:00.000Z
BTC_USDT |     0.00015 | 1639008000000 | 2021-12-09T00:00:00.000Z
BTC_USDT |     0.00005 | 1639036800000 | 2021-12-09T08:00:00.000Z
BTC_USDT |     0.00015 | 1639065600000 | 2021-12-09T16:00:00.000Z
BTC_USDT |     0.00015 | 1639094400000 | 2021-12-10T00:00:00.000Z
BTC_USDT |     0.00015 | 1639123200000 | 2021-12-10T08:00:00.000Z
BTC_USDT |     0.00015 | 1639152000000 | 2021-12-10T16:00:00.000Z
BTC_USDT |     0.00015 | 1639180800000 | 2021-12-11T00:00:00.000Z
20 objects
```

### CODE REVIEW

**Comment by frosty00** [comment]
we don't need to worry about paging, a user can pass those via the implicit params

**Comment by frosty00** [comment]
page_size should be `limit` as you say

**Comment by samgermain** [comment]
> we don't need to worry about paging, a user can pass those via the implicit params

It's required so we'll at least need 

```
const pageNum = this.safeInteger (params, 'page_num');
if (pageNum === undefined) {
    throw new ArgumentsRequired (this.id + ' fetchFundingRate history requires params.page_num');
}
```

Or should we set `const pageNum = this.safeInteger (params, 'page_num');` with now Error thrown?

File: js/mexc.js [hunk 1]
```diff
@@ -2094,4 +2095,66 @@ module.exports = class mexc extends Exchange {
...
+        // Gets a history of funding rates with their timestamps
+        //  (param) symbol: Future currency pair
+        //  (param) limit: not used by mexc
+        //  (param) since: not used by mexc
+        //  (param) params: Object containing more params for the request
+        //  return: [{symbol, fundingRate, timestamp, dateTime}]
+        //
+        if (symbol === undefined) {
+            throw new ArgumentsRequired (this.id + ' fetchFundingRateHistory() requires a symbol argument');
+        }
```
Comment (samgermain) [inline js/mexc.js:2110]: you should also have a check like this for limit, because `mexc` requires it

**Comment by kroitor** [comment]
@samgermain 

> ```
> const pageNum = this.safeInteger (params, 'page_num');
> if (pageNum === undefined) {
>     throw new ArgumentsRequired (this.id + ' fetchFundingRate history requires params.page_num');

Sidenote – we should always use standardized exception messages:

```JavaScrpipt
     throw new ArgumentsRequired (this.id + ' fetchFundingRateHistory() requires a page_num parameter');
```

In this case, however, it is not requires, cause we have a `limit` argument and if the `limit` argument is `undefined` we should default that to the exchange's default value or to the exchange's max value.

**Comment by Dan-krm** [comment]
I did a check to see if limit is undefined then set it to the page_size default value. Let me know if there's anything else I should change for this method.

**Comment by frosty00** [comment]
@Dan-krm we don't set the limit to the default if it the limit is undefined. Instead we should just omit the `page_size` param from the request in this case.

**Comment by Dan-krm** [comment]
> @Dan-krm we don't set the limit to the default if it the limit is undefined. Instead we should just omit the `page_size` param from the request in this case.

Okay, can I have a bit more of an explanation of how that works, are you saying I can just leave it how it was before, or is there a specific way to omit a param from the request if it's undefined?

**Comment by kroitor** [comment]
> Okay, can I have a bit more of an explanation of how that works, are you saying I can just leave it how it was before, or is there a specific way to omit a param from the request if it's undefined?

You can just leave it out.

**Comment by kroitor** [comment]
In general, all unified method signatures (names, return types and arguments) must be in line across all exchanges in the lib. With all unified methods, we do the following template:

```JavaScript
// do not confuse arguments and params
// these we call "arguments"         ↓                    ↓                          ↓
//                                                                                                                                             ↓ and these are "params"
async fetchSomething (symbol = undefined, since = undefined, limit = undefined, params = {})
    // if the unified method signature does not require a symbol
    // but the exchange requires it, then we throw early
    if (symbol === undefined) {
        throw new ArgumentsRequired (this.id + ' fetchSomething() requires a symbol argument')
    }
    await this.loadMarkes ();
    const market = this.market (symbol);
    // we document the request in comments here    
    const request = {
        'symbol': market['id'], // required
        // 'page_size': limit, // optional, we document it here in the comment
        // 'foo': 'bar', // some other optional parameter – we document them all in comments
    };
    if (limit !== undefined) {
        request['page_size'] = limit; // only set it if the limit argument is defined
    }
    // always extend the request with user-overrided params ↓
    const response = await this.someImplicitEnpoint (this.extend (request, params));
    // we document the response in comments here
    // ...
```

**Comment by kroitor** [comment]
Another template, when the symbol is not required by the exchange:

```JavaScript
// do not confuse arguments and params
// these we call "arguments"         ↓                    ↓                          ↓
//                                                                                                                                             ↓ and these are "params"
async fetchSomething (symbol = undefined, since = undefined, limit = undefined, params = {})
    await this.loadMarkes ();
    let market = undefined;
    // we document the request in comments here    
    const request = {
        // 'symbol': market['id'], // optional
        // 'page_size': limit, // optional, we document it here in the comment
        // 'foo': 'bar', // some other optional parameter – we document them all in comments
    };
     if (symbol !== undefined) {
        market = this.market (symbol);
        request['symbol'] = market['id']; // only set it if the symbol argument is defined
    }
    if (limit !== undefined) {
        request['page_size'] = limit; // only set it if the limit argument is defined
    }
    // always extend the request with user-overrided params ↓
    const response = await this.someImplicitEnpoint (this.extend (request, params));
    // we document the response in comments here
    // ...
```

**Comment by Dan-krm** [comment]
Thanks, those explanations are helpful. I think this method is okay now, let me know if there's anything else I should be aware of

File: js/mexc.js (hunk 1 above)
Comment (kroitor) [reply js/mexc.js:2110]: @samgermain looks like it works without `page_size` and without `page_num`:

- https://contract.mexc.com/api/v1/contract/funding_rate/history?symbol=BTC_USDT

But not without a symbol:

- https://contract.mexc.com/api/v1/contract/funding_rate/history

**Review by codywalker77 — COMMENTED** [review]
Q

**Pull request merged**